File: packages/server/src/domains/flags/service.ts
import type { Flag, CreateFlagInput, UpdateFlagInput } from '@flick/shared';
import * as repo from './repo.js';
import { NotFoundError, ConflictError } from '../../shared/errors.js';

export async function listFlags(
  projectId: string,
  opts?: { archived?: boolean; tags?: string[] },
): Promise<Flag[]> {
  return repo.findByProject(projectId, opts);
}

export async function getFlag(id: string): Promise<Flag> {
  const flag = await repo.findById(id);
  if (!flag) throw new NotFoundError('Flag', id);
  return flag;
}

export async function getFlagByKey(projectId: string, key: string): Promise<Flag> {
  const flag = await repo.findByKey(projectId, key);
  if (!flag) throw new NotFoundError('Flag', key);
  return flag;
}

export async function createFlag(projectId: string, input: CreateFlagInput): Promise<Flag> {
  const existing = await repo.findByKey(projectId, input.key);
  if (existing) throw new ConflictError(`Flag with key '${input.key}' already exists`);
  return repo.create(projectId, input);
}

export async function updateFlag(id: string, input: UpdateFlagInput): Promise<Flag> {
  const flag = await repo.update(id, input);
  if (!flag) throw new NotFoundError('Flag', id);
  return flag;
}

export async function archiveFlag(id: string): Promise<Flag> {
  const flag = await repo.archive(id);
  if (!flag) throw new NotFoundError('Flag', id);
  return flag;
}

export async function restoreFlag(id: string): Promise<Flag> {
  const flag = await repo.restore(id);
  if (!flag) throw new NotFoundError('Flag', id);
  return flag;
}

export async function deleteFlag(id: string): Promise<void> {
  const deleted = await repo.remove(id);
  if (!deleted) throw new NotFoundError('Flag', id);
}

File: packages/server-rust/src/domains/flags/service.rs
use sqlx::PgPool;
use uuid::Uuid;

use crate::error::AppError;

use super::repo::{self, FlagRow};

pub async fn list_flags(
    db: &PgPool,
    project_id: Uuid,
    archived: Option<bool>,
    tags: Option<Vec<String>>,
) -> Result<Vec<FlagRow>, AppError> {
    repo::find_by_project(db, project_id, archived, tags)
        .await
        .map_err(AppError::from)
}

pub async fn get_flag(db: &PgPool, id: Uuid) -> Result<FlagRow, AppError> {
    repo::find_by_id(db, id)
        .await?
        .ok_or_else(|| AppError::not_found("Flag", Some(&id.to_string())))
}

pub struct CreateFlagInput<'a> {
    pub key: &'a str,
    pub name: &'a str,
    pub description: &'a str,
    pub gate_type: &'a str,
    pub tags: &'a [String],
}

pub async fn create_flag(
    db: &PgPool,
    project_id: Uuid,
    input: CreateFlagInput<'_>,
) -> Result<FlagRow, AppError> {
    // Check for key conflict within project
    if repo::find_by_key(db, project_id, input.key).await?.is_some() {
        return Err(AppError::conflict(format!(
            "Flag with key '{}' already exists in this project",
            input.key
        )));
    }

    repo::create(
        db,
        project_id,
        input.key,
        input.name,
        input.description,
        input.gate_type,
        input.tags,
    )
    .await
}

pub struct UpdateFlagInput {
    pub name: Option<String>,
    pub description: Option<String>,
    pub tags: Option<Vec<String>>,
}

pub async fn update_flag(
    db: &PgPool,
    id: Uuid,
    input: UpdateFlagInput,
) -> Result<FlagRow, AppError> {
    repo::update(
        db,
        id,
        input.name.as_deref(),
        input.description.as_deref(),
        input.tags.as_deref(),
    )
    .await?
    .ok_or_else(|| AppError::not_found("Flag", Some(&id.to_string())))
}

pub async fn archive_flag(db: &PgPool, id: Uuid) -> Result<FlagRow, AppError> {
    repo::archive(db, id)
        .await?
        .ok_or_else(|| AppError::not_found("Flag", Some(&id.to_string())))
}

pub async fn restore_flag(db: &PgPool, id: Uuid) -> Result<FlagRow, AppError> {
    repo::restore(db, id)
        .await?
        .ok_or_else(|| AppError::not_found("Flag", Some(&id.to_string())))
}

pub async fn delete_flag(db: &PgPool, id: Uuid) -> Result<Uuid, AppError> {
    repo::remove(db, id)
        .await?
        .ok_or_else(|| AppError::not_found("Flag", Some(&id.to_string())))
}
